extra/yolov3.py

```diff
@@ -0,0 +1,626 @@
+      try:
+        ious = bbox_iou(np.expand_dims(image_pred_class[i], axis=0), image_pred_class[i+1:])
+      except ValueError:
+        break
+  
+      except IndexError:
+        break
+  
+      #Zero out all the detections that have IoU > treshhold
+      iou_mask = np.expand_dims((ious < nms_conf), axis=1)
+      # iou_mask = (ious < nms_conf).float().unsqueeze(1)
+      image_pred_class[i+1:] *= iou_mask
+  
+      #Remove the non-zero entries
+      non_zero_ind = np.squeeze(np.nonzero(image_pred_class[:,4]))
+      image_pred_class = np.reshape(image_pred_class[non_zero_ind], (-1, 7))    
+
+    batch_ind = np.array([[0]])
+    seq = (batch_ind, image_pred_class)
+    
+    if not write:
+      output = np.concatenate(seq, 1)
+      write = True
+    else:
+      out = np.concatenate(seq, axis=1)
+      output = np.concatenate((output,out))
+  try:
+    return output
+  except:
+    return 0
+
+def imresize(img, w, h):
+  return np.array(Image.fromarray(img).resize((w, h)))
+
+def resize(img, inp_dim=(608, 608)):
+  img_w, img_h = img.shape[1], img.shape[0]
+  w, h = inp_dim
+  new_w = int(img_w * min(w/img_w, h/img_h))
+  new_h = int(img_h * min(w/img_w, h/img_h))
+  resized_image = cv2.resize(img, (new_w,new_h), interpolation = cv2.INTER_CUBIC)
+  
+  canvas = np.full((inp_dim[1], inp_dim[0], 3), 128)
+  canvas[(h-new_h)//2:(h-new_h)//2 + new_h,(w-new_w)//2:(w-new_w)//2 + new_w,  :] = resized_image
+  
+  return canvas
+
+def infer(model, img):
+  img = np.array(img)
+  img = imresize(img, 608, 608)
+  # img = resize(img)
+  img = img[:,:,::-1].transpose((2,0,1))
+  img = img[np.newaxis,:,:,:]/255.0
+
+  prediction = model.forward(Tensor(img))
+  return prediction
+
+
+def parse_cfg(cfg):
+  # Return a list of blocks
+  lines = cfg.decode("utf-8").split('\n')
+  lines = [x for x in lines if len(x) > 0]
+  lines = [x for x in lines if x[0] != '#']
+  lines = [x.rstrip().lstrip() for x in lines]
+
+  block = {}
+  blocks = []
+
+  for line in lines:
+    if line[0] == "[":
+      if len(block) != 0:
+        blocks.append(block)
+        block = {}
+      block["type"] = line[1:-1].rstrip()
+    else:
+      key,value = line.split("=")
+      block[key.rstrip()] = value.lstrip()
+  blocks.append(block)
+
+  return blocks
+
+# TODO: Speed up this function, avoid copying stuff from GPU to CPU
+def predict_transform(prediction, inp_dim, anchors, num_classes):
+  batch_size = prediction.shape[0]
+  stride = inp_dim // prediction.shape[2]
+  grid_size = inp_dim // stride
+  bbox_attrs = 5 + num_classes
+  num_anchors = len(anchors)
+  
+  prediction = prediction.reshape(shape=(batch_size, bbox_attrs*num_anchors, grid_size*grid_size))
+  # Original PyTorch: transpose(1, 2) -> For some reason numpy.transpose order has to be reversed?
+  prediction = prediction.transpose(order=(0, 2, 1))
+  prediction = prediction.reshape(shape=(batch_size, grid_size*grid_size*num_anchors, bbox_attrs))
+  
+  prediction_cpu = prediction.cpu().data
```

Comment:
@geohot Any idea why this line is taking ~1s to run? Im close to getting decent results on GPU, this is the only line left taking a long time… 

When i run the model with bigger input size `width=608` the line takes approximately one second to run, however, when I run it with `width=416` (smaller input size, .cpu() should return faster) the line takes 5 seconds!! There has to be something underlying here, to point out a reason why a larger tensor takes less time to put on the CPU...

Once this is figured out, we can actually achieve some good results on webcam using GPU!